### PR TITLE
Implement local_config_platform in @platforms

### PR DESCRIPTION
**Pull request by Wyverald**
Per design doc: https://docs.google.com/document/d/1g5JAAOfLsvQKBGqzSLFp1hIYFoQsgOslsjaIGV6P7Tk/edit

plan:

1. Merge this PR and push a new release.
2. Make `bazel_tools` depend on the new version of `platforms`.
3. Make `--host_platform` default to `@bazel_tools//:host_platform`, which is an alias to `@platforms//host`.
    * This is potentially problematic. Do aliases work with platforms?
4. Update [`LocalConfigPlatformFunction.java`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/repository/LocalConfigPlatformFunction.java) in Bazel source to make it produce a `@local_config_platform` that forwards everything here:
    * `@local_config_platform//:host` becomes an alias to `@platforms//host`.
    * `@local_config_platform//:constraints.bzl` simply re-exports the `HOST_CONSTRAINTS` list from `@platforms//host:constraints.bzl`.
5. Add an incompatible flag that removes the built-in module `local_config_platform`.
6. The above is cherry-pickable back to Bazel 7.x. We can start advertising to people to stop using `@local_config_platform` directly, and use the stuff in `@platforms//host` instead.
7. Flip the incompatible flag in Bazel 8.0, with the intention to completely remove `local_config_platform` (both the repo rule and the built-in module) in Bazel 9.0.
8. Implement the custom constraint injection stuff here as we see fit. (see https://github.com/fmeum/host_platform/tree/main)
9. Profit??

### CODE REVIEW

**Comment by Wyverald** [comment]
cc @fmeum @katre @brentleyjones

@katre specifically, do aliases work with platforms? If they don't, the "canonical" host platform will need to be `@bazel_tools//:host_platform` instead of `@platforms//host` (because default flag values can only be stuff inside built-in modules like `@bazel_tools` (and `@local_config_platform` today)) and the migration path is a bit trickier (we'd have to ask people to actually stop using `@local_config_platform//:host` and switch to `@bazel_tools//:host_platform` instead).

**Comment by katre** [comment]
For future reference: context here is https://github.com/bazelbuild/bazel/issues/8766 and https://bazelbuild.slack.com/archives/CA31HN1T3/p1709673438139349

@Wyverald: There are reasons I have been asking for an actual design for this feature from the beginning of https://github.com/bazelbuild/bazel/issues/8766. We very deliberately have resisted putting complex logic in the `@platforms` repository, because it is foundational to so many other repos (including Bazel itself), and so anything that we add here has a lot of follow-on effects. I'm not opposed to it, but we need to plan for it.

I'd like to, once again, ask people to step back from the fun part of writing code and move to the less fun part of writing and discussing a design document.

**Comment by Wyverald** [comment]
Design doc: https://docs.google.com/document/d/1g5JAAOfLsvQKBGqzSLFp1hIYFoQsgOslsjaIGV6P7Tk/edit

**Comment by Wyverald** [comment]
The test I added is a bit ugly, but it works... except on Windows. Not sure what's going on there...

File: host/BUILD.bazel
Comment (katre) [inline host/BUILD.bazel:1]: There should be an initial comment explaining the purpose of these files.

File: tests/BUILD
Comment (katre) [inline tests/BUILD:1]: As things are currently set up, will these test files be included in the release? We should avoid that if posible.

**Review by katre — CHANGES_REQUESTED** [review]

**Comment by katre** [comment]
Added @aranguyen as an FYI

**Comment by katre** [comment]
I don't think the `sh_test` can execute on windows, do our images have bash support anymore?

**Comment by meteorcloudy** [comment]
> I don't think the sh_test can execute on windows, do our images have bash support anymore?

`sh_test` should work on Windows and we do have MSYS2 available on CI Windows VMs. 

Believe it or not, the shell test was failing because the Windows launcher wasn't built for the shell test, and that's because the [isExecutedOnWindows](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/RuleContext.java;drc=17e855c959edd42f241e2767f7e434d3f452ac34;bpv=1;bpt=1;l=1446?gsn=isExecutedOnWindows&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fgithub.com%2Fbazelbuild%2Fbazel%3Flang%3Djava%3Fpath%3Dcom.google.devtools.build.lib.analysis.RuleContext%23a94f0a8dde309bfecd095864efded204adaaba7117d53326d323bce6a9a5fa60) function is failing when the root module's name is `platforms`!

**Comment by meteorcloudy** [comment]
When the root module's name is not `platforms`

image


when it is `platforms`:

image

File: host/BUILD.bazel
Comment (Wyverald) [reply host/BUILD.bazel:1]: added a small comment (similar to what's in cpu/BUILD)

File: tests/BUILD
Comment (Wyverald) [reply tests/BUILD:1]: no they won't. but I also realized that the new `host` directory is also not included 😅 fixed.

**Review by katre — APPROVED** [review]

**Comment by aignas** [comment]
I am trying to use this and I want to build docs with `stardoc` for a library that uses this. Any tips on how to construct the `bzl_library`? I get
```
$ bazel build //docs/sphinx/...
ERROR: /Users/ignas.anikevicius/src/github/aignas/rules_python/docs/sphinx/BUILD.bazel:75:16: in deps attribute of starlark_doc_extract rule //docs/sphinx:_bzl_api_docs_api_extensions_pip.md_stardoc: missing bzl_library targets for Starlark module(s) @@platforms~host_platform~host_platform//:constraints.bzl. Since this rule was created by the macro 'sphinx_stardocs', the error might have been caused by the macro implementation
ERROR: /Users/ignas.anikevicius/src/github/aignas/rules_python/docs/sphinx/BUILD.bazel:75:16: Analysis of target '//docs/sphinx:_bzl_api_docs_api_extensions_pip.md_stardoc' failed
Use --verbose_failures to see the command lines of failed build steps.
ERROR: Analysis of target '//docs/sphinx:_bzl_api_docs_api_extensions_pip.md' failed; build aborted: Analysis failed
INFO: Elapsed time: 0.220s, Critical Path: 0.00s
INFO: 0 processes.
ERROR: Build did NOT complete successfully
```

The code is in bazelbuild/rules_python#1827.

**Pull request merged**